{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "electronic-installation",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import scipy.io\n",
    "import numpy as np\n",
    "from tqdm.notebook import tqdm, trange\n",
    "import pickle\n",
    "import pandas as pd\n",
    "\n",
    "from iGEM import iGEM\n",
    "from iGEM_fixed_w import iGEM_fixed_w\n",
    "\n",
    "import seaborn as sns; sns.set(color_codes=True)\n",
    "\n",
    "from scipy.stats.stats import pearsonr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "extended-breed",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_topic(beta, gene_name_path, save_name, top_gene_num):\n",
    "    phi = beta\n",
    "    phi = pd.DataFrame(phi)\n",
    "    try:\n",
    "        if gene_name_path == 'mRNA':\n",
    "            gene_name = pd.DataFrame(dataset['mRNA_ref'])\n",
    "        elif gene_name_path == 'methylation':\n",
    "            gene_name = pd.DataFrame(dataset['methylation_ref']).astype(str)\n",
    "        elif gene_name_path == 'geneset':\n",
    "            gene_name = pd.DataFrame(dataset['geneset_ref'])\n",
    "        else:\n",
    "            try:\n",
    "                gene_name = pd.read_csv(gene_name_path, sep = ',', dtype='str', header = None)\n",
    "            except:\n",
    "                gene_name = gene_name_path\n",
    "    except:\n",
    "        gene_name = gene_name_path\n",
    "    gene_name.columns = ['Name']\n",
    "    phi.columns = [f'{i}' for i in range(len(phi.columns))]\n",
    "    phi = phi.astype(float)\n",
    "    phi = pd.concat([gene_name['Name'], phi], axis = 1)\n",
    "    dataset_lst = [phi]\n",
    "    final_phi_lst = []\n",
    "    for data in dataset_lst:\n",
    "        final_phi = data.copy()\n",
    "        final_phi = final_phi.sort_values(by=['0'], ascending = False)\n",
    "        final_phi = final_phi.iloc[0:top_gene_num, :]\n",
    "        for i in trange(len(data.columns)-2):\n",
    "            tmp_phi = data.sort_values(by=[f'{i+1}'], ascending = False)\n",
    "            final_phi = pd.concat([final_phi, tmp_phi.iloc[0:top_gene_num, :]])\n",
    "        final_phi_lst.append(final_phi)\n",
    "    for i in range(1):\n",
    "        data = final_phi_lst[i]\n",
    "        data = data.set_index('Name')\n",
    "        #for names that are too long\n",
    "        data.index = [name[:40]+\"...\" if len(name)>40 else name for name in list(data.index)]\n",
    "        data.columns = [int(i)+1 for i in list(data.columns)]\n",
    "        sns.set(font_scale=1.4)\n",
    "        #dendrogram_ratio for controlling space around\n",
    "        g = sns.clustermap(data, figsize=(20, 3*params_dict['k']*top_gene_num/10), row_cluster = False, col_cluster = False, cmap = \"RdBu_r\", \n",
    "                           xticklabels=True, yticklabels=True, center=0, cbar_pos=(0, .2, .03, .4),\n",
    "                          dendrogram_ratio=(0.075,0.001))\n",
    "        g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_ymajorticklabels(), fontsize = 18)\n",
    "        g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xmajorticklabels(), fontsize = 18)\n",
    "        g.ax_heatmap.hlines([(i+1)*top_gene_num for i in range(params_dict['k'])], *g.ax_heatmap.get_xlim())\n",
    "        #g.savefig(save_name + '_feature_topic.pdf')\n",
    "        \n",
    "def normalize_sum(x):\n",
    "    x_normed = x / x.sum(0, keepdim=True)[0]\n",
    "    x_normed[x_normed != x_normed] = 0\n",
    "    return x_normed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "timely-necessity",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = f'processed_dataset.pkl'\n",
    "with open(data_path, 'rb') as handle:\n",
    "    dataset = pickle.load(handle)\n",
    "\n",
    "#mouse rnaseq single section\n",
    "#params_dict = set_model_param()\n",
    "params_dict = {}\n",
    "params_dict['rho1'] = dataset['rho1'].to(torch.float)\n",
    "#params_dict['rho2'] = mouse_omic['rho2'].to(torch.float)\n",
    "true_rho = params_dict['rho1']\n",
    "X1 = dataset['mRNA'].to(torch.float)\n",
    "X2 = dataset['methylation'].to(torch.float)\n",
    "\n",
    "mRNA = dataset['mRNA'].detach().numpy()\n",
    "met = dataset['methylation'].detach().numpy()\n",
    "\n",
    "#define mode flags\n",
    "params_dict['omic_num'] = 2\n",
    "params_dict['use_alpha'] = [True, False, False]\n",
    "params_dict['use_poisson'] = [False, False, False]\n",
    "\n",
    "params_dict['k'] = 10\n",
    "params_dict['max_iter'] = 3000\n",
    "params_dict['tol'] = 1e-8\n",
    "params_dict['pi_aa'] = 0\n",
    "params_dict['pi_ab'] = 0.01\n",
    "params_dict['pi_bb'] = 0.01\n",
    "params_dict['pi_xr'] = 2e-6\n",
    "params_dict['pi_xc'] = 0.2\n",
    "params_dict['xc_h2_coef'] = 0.01\n",
    "params_dict['xc_alpha1_coef'] = 0.05\n",
    "params_dict['prev_model'] = None\n",
    "\n",
    "#mode week8 NRES tf-gene 0.4/0.4\n",
    "aa_path = 'SNMNMF/rnaseq_interaction_sparse_matrix_deseq1639.pkl'\n",
    "aa_path = None\n",
    "#ab_path = 'SNMNMF/bader_met_embedding_omic_data_0610.pkl'\n",
    "ab_path = None\n",
    "bb_path = 'SNMNMF/rnaseq_interaction_sparse_matrix_deseq5000.pkl'\n",
    "bb_path = None\n",
    "if aa_path == None:\n",
    "    aa = torch.zeros(X1.shape[1], X1.shape[1])\n",
    "else:\n",
    "    with open(aa_path, 'rb') as handle:\n",
    "        aa = pickle.load(handle)\n",
    "        aa = aa.to_dense()\n",
    "if ab_path == None:\n",
    "    ab = torch.zeros(X1.shape[1], X2.shape[1])\n",
    "else:\n",
    "    with open(ab_path, 'rb') as handle:\n",
    "        ab = pickle.load(handle)\n",
    "        ab = ab.to_dense()\n",
    "if bb_path == None:\n",
    "    bb = torch.zeros(X2.shape[1], X2.shape[1])\n",
    "else:\n",
    "    with open(bb_path, 'rb') as handle:\n",
    "        bb = pickle.load(handle)\n",
    "        bb = bb.to_dense()\n",
    "model = iGEM(X1,X2,aa,ab,bb, params_dict)\n",
    "model.train()\n",
    "gene_topic(normalize_sum(model.H1.t().detach()).numpy(), 'mRNA', 'test', 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "muslim-british",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_topic(normalize_sum(model.alpha1.t().detach()).numpy(), 'geneset', 'test', 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "automatic-hamilton",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_topic(normalize_sum(model.H1.t().detach()).numpy(), 'mRNA', 'test', 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lesser-audience",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_topic(normalize_sum(model.alpha1.t().detach()).numpy(), 'geneset', 'test', 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sorted-convenience",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "above-nickel",
   "metadata": {},
   "outputs": [],
   "source": [
    "#test section\n",
    "params_dict = {}\n",
    "source_w = torch.rand(111, 10)\n",
    "source_rho = np.mod(np.random.permutation(100*100).reshape(100,100),2)\n",
    "source_rho = scipy.sparse.random(100,100, density=0.1, random_state=2020, data_rvs=np.ones)\n",
    "source_rho = source_rho.A\n",
    "source_rho = torch.from_numpy(source_rho).float()\n",
    "params_dict['rho1'] = source_rho\n",
    "source_alpha = scipy.sparse.random(source_w.shape[1], source_rho.shape[0], density=0.1, random_state=2020)\n",
    "source_alpha = source_alpha.A\n",
    "source_alpha = torch.from_numpy(source_alpha).float()\n",
    "source_X = torch.mm(torch.mm(source_w, source_alpha), source_rho)\n",
    "X1 = source_X\n",
    "X2 = source_X\n",
    "\n",
    "fixed_model_w = source_w\n",
    "\n",
    "#define mode flags\n",
    "params_dict['omic_num'] = 1\n",
    "params_dict['use_alpha'] = [True, False, True]\n",
    "params_dict['use_poisson'] = [False, False, False]\n",
    "\n",
    "params_dict['k'] = 10\n",
    "params_dict['max_iter'] = 500\n",
    "params_dict['tol'] = 1e-8\n",
    "params_dict['pi_aa'] = 0\n",
    "params_dict['pi_ab'] = 0.01\n",
    "params_dict['pi_bb'] = 0.01\n",
    "params_dict['pi_xr'] = 2e-6\n",
    "params_dict['pi_xc'] = 0.2\n",
    "params_dict['xc_h2_coef'] = 0.01\n",
    "params_dict['xc_alpha1_coef'] = 0.05\n",
    "\n",
    "params_dict['prev_model'] = None\n",
    "\n",
    "aa_path = None\n",
    "ab_path = None\n",
    "bb_path = None\n",
    "if aa_path == None:\n",
    "    aa = torch.zeros(X1.shape[1], X1.shape[1])\n",
    "else:\n",
    "    with open(aa_path, 'rb') as handle:\n",
    "        aa = pickle.load(handle)\n",
    "        aa = aa.to_dense()\n",
    "if ab_path == None:\n",
    "    ab = torch.zeros(X1.shape[1], X2.shape[1])\n",
    "else:\n",
    "    with open(ab_path, 'rb') as handle:\n",
    "        ab = pickle.load(handle)\n",
    "        ab = ab.to_dense()\n",
    "if bb_path == None:\n",
    "    bb = torch.zeros(X2.shape[1], X2.shape[1])\n",
    "else:\n",
    "    with open(bb_path, 'rb') as handle:\n",
    "        bb = pickle.load(handle)\n",
    "        bb = bb.to_dense()\n",
    "\n",
    "model = iGEM_fixed_w(X1,X2,aa,ab,bb, params_dict, fixed_model_w)\n",
    "model.train()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "minus-period",
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha1_source = source_alpha.t().detach().numpy()\n",
    "alpha1_true = model.alpha1.t().detach().numpy()\n",
    "g_source = sns.clustermap(alpha1_source, figsize=(6, 8), row_cluster = False, col_cluster = False, cmap = \"RdBu_r\", xticklabels=True, yticklabels=False, center=0)\n",
    "g_true = sns.clustermap(alpha1_true, figsize=(6, 8), row_cluster = False, col_cluster = False, cmap = \"RdBu_r\", xticklabels=True, yticklabels=False, center=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "least-stopping",
   "metadata": {},
   "outputs": [],
   "source": [
    "#derived score correlates well with source score\n",
    "source_matrix = alpha1_source\n",
    "true_matrix = alpha1_true\n",
    "score = []\n",
    "for i in trange(source_matrix.shape[1]):\n",
    "    for j in range(true_matrix.shape[1]):\n",
    "        stat, p = pearsonr(source_matrix[:,i], true_matrix[:,j])\n",
    "        score.append([i, j, stat, p])\n",
    "score = pd.DataFrame(score, columns = ['topic1', 'topic2', 'coef', 'pval'])\n",
    "score = score.sort_values(by=['coef'], ascending=False).reset_index(drop=True)\n",
    "score.iloc[:15]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lyric-dominican",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
